### ISSUE TITLE
EQ2 design: the equanimity rule reduces to a fixed weight where units agree — pre-register it as a units-fixer, with the Ω drift predicted (R3, R4, R7

### ISSUE THREAD

**Issue by alexsabine** [alexsabine/ashes_crr#20]
Study EQ2 — the equanimity rule as a tuning-free step for online EWC

Prompt for the agent running this study. Paste it whole. It is written to be executed in alexsabine/ashes_crr under CLAUDE.md; the rules there (R1–R13) override anything in this prompt. If any instruction below conflicts with a rule, stop and say so.

0. Before anything else
Read CLAUDE.md in full, then [theory/CRR.md](https://github.com/alexsabine/ashes_crr/blob/claude/relaxed-hypatia-m8ikp5/theory/CRR.md) in full, then [reports/eqx.md](https://github.com/alexsabine/ashes_crr/blob/claude/relaxed-hypatia-m8ikp5/reports/eqx.md) and ledger rows EQX-1 to EQX-5.
Log this prompt verbatim in [notebook/PROMPT_LOG.md](https://github.com/alexsabine/ashes_crr/blob/claude/relaxed-hypatia-m8ikp5/notebook/PROMPT_LOG.md) with a UTC timestamp (R13).
Study id is eq2. Use the command order in CLAUDE.md §8 exactly: gate, prereg, hash, OpenTimestamps anchor, signed tag, and only then data.
Every number in the prereg thresholds below is fixed now. Nothing in this prompt may be changed after the tag exists (R3). If you must change it, the study is void and gets a new id.
1. The claim, in plain words and then precisely

Plain. A model learning something new is like a wagon pulled by two kids. The New kid pulls toward the new task, the Old kid toward what the model already knows. Every continual-learning method adds the two pulls with a hand-set weight on the Old kid, and that weight has to be retuned for every dataset because the two pulls are measured in different units. The equanimity rule says: at every step, measure how hard each kid is pulling right now, and make the Old kid's pull the same length as the New kid's. Ω = 1 means "the same length". The rule is not a weight chosen once. It is a step in the Old direction whose length is always the length of the New step.

Precise. With g_present the gradient of the current-task loss and g_past the gradient of the past term (both over all parameters), the update direction is

g_present + w · g_past,   w = Ω · ‖ĝ_present‖ / ‖ĝ_past‖,   Ω = 1

where ĝ is an exponential moving average of the gradient vector (not of its norm), the norm is Euclidean, and w is capped. Algebraically, w · g_past = Ω · ‖ĝpresent‖ · (g_past / ‖ĝpast‖) up to the smoothing: the past term's direction with the present term's step length.

Where it should work, and why. Study EQX showed that when the past term is the same loss as the present term (plain replay), the rule reduces to a constant near 1 and adds nothing. The claim for EQ2 is that the rule is useful exactly when the past term is the past task's loss or its Fisher-curvature (Laplace) approximation, which is online EWC's penalty Σ F_i (θi − θ*i)². There the penalty gradient is near zero at the anchor and grows with displacement, so a fixed λ that is stable far from the anchor is too weak near it, and a fixed λ that holds near it diverges far from it; a step of fixed length in the penalty direction is stable in both places. Nothing in this claim depends on the units of the penalty, so it removes EWC's λ, the most dataset-dependent hyperparameter in the field.

Where it must not work. Same-units replay (ER-sum): redundant. Heuristic constraints that are not a loss on the past task (DER++ logit matching, LwF distillation): the correct strength of those terms is a small fraction of the present's, so "same length" over-regularises. Importance penalties whose weights are not a curvature of the past loss (SI, MAS): the penalty direction is not a descent direction on the old task. These are the controls. If any of them passes, the mechanism is wrong and the study says so.

2. Exact definitions (write these into the prereg verbatim)
g_present: gradient of the cross-entropy on the current stream batch, all parameters, flattened.
g_past, per method: EWC-online: gradient of the penalty Σ F_i (θi − θ*i)² with λ removed. ER-sum: gradient of the cross-entropy on the replay batch (a separate batch mean, summed with the stream batch mean; not Mammoth's concatenated single mean). DER++: gradient of the logit-matching term with α removed. LwF: gradient of the distillation term with its weight removed. SI, MAS: gradient of the importance penalty with its weight removed.
Estimator, fixed at hash time: ratio = ema of the gradient vectors, smooth = 0.9, Euclidean norm (chosen on EQX, where Euclidean beat Fisher on 3/3 datasets; this prereg names that as the source), cap = 1e4, floor on the denominator 1e-12. The cap is a named parameter and is swept in the sensitivity table; the registered verdict uses cap = 1e4.
Score: Mammoth's final Class-IL mean accuracy over all tasks, per seed, seeds 0–4. Aggregation is the mean over five seeds. Per-seed values are reported next to every mean (R6).
Resolvable step, per carrier: step = max(1.0 pt, 2 × SE) where SE is the standard error of the tuned-λ EWC accuracy over the five seeds. Written into the ledger row. No threshold is finer than the step (R5).
Tuned λ, per carrier and method: the grid value with the highest five-seed mean. λ is tuned on the same seeds it is scored on. This is in-sample for λ and therefore biased in favour of the fixed baseline, which is the direction we want.
3. Phase A: the gate (before the prereg, R4)

Extend [src/crr/surrogates/gate.py](https://github.com/alexsabine/ashes_crr/blob/claude/relaxed-hypatia-m8ikp5/src/crr/surrogates/gate.py) with gate_EQ2. Add to MUST_PASS a convex learner whose past term is the exact quadratic (Laplace) penalty of a past quadratic loss with parameter scale mismatched 16× against the present term; add to MUST_FAIL (a) the existing same-units replay learner S-R and (b) a learner whose past term is a constraint whose tuned weight is a small fraction of the present gradient's norm. The gate statistic is the EQ2 statistic: "rule at Ω = 1 not behind the tuned fixed weight by a step". Commit [prereg/eq2/gate_EQ2.txt](https://github.com/alexsabine/ashes_crr/blob/claude/relaxed-hypatia-m8ikp5/prereg/eq2/gate_EQ2.txt) and [runs/phaseA/gate_EQ2.txt](https://github.com/alexsabine/ashes_crr/blob/claude/relaxed-hypatia-m8ikp5/runs/phaseA/gate_EQ2.txt); the gate must read OPEN before the prereg is written. If it does not open, stop and report (R12).

4. Pre-registered rows (prereg order; each becomes a ledger row)
EQ2-0 Precondition (decidability). For EWC-online, tuned λ across the three carriers spans ≥ 10× (max/min). If not, the row reads "not decidable: tuned λ did not move" and rows EQ2-1 and EQ2-2 are reported without a verdict.
EQ2-1 H-EQ2 (the claim). On every carrier, mean(rule at Ω = 1) − mean(tuned λ) > −step, AND the best single λ across carriers is behind the tuned λ by ≥ step on at least one carrier. PASS requires both. FAIL if the rule is behind by ≥ step on any carrier, or if a single λ transfers. Two-sided reporting; the verdict is one-sided by design (the rule need only not lose).
EQ2-2 Reduction test. Fixed w = per-carrier median of the derived w from the Ω = 1 run. If mean(fixed at median) is within a step of mean(rule) on every carrier, the row reads "reduces to a per-carrier constant". If it is behind by ≥ step or diverges on any carrier, the row reads "does not reduce: normalised-gradient method". Either outcome is reported, neither is a failure.
EQ2-3 Control, same units (ER-sum). Must FAIL to help: mean(rule at Ω = 1) − mean(best fixed w on the grid {0.5, 1, 2, 4}) < step on every carrier, and the reduction test reads "reduces". If the rule beats the best fixed w by ≥ step on any carrier, the control is violated.
EQ2-4 Control, constraint (DER++ and LwF). Must FAIL to land: for each of the two methods, on every carrier, the rule at its best Ω on the grid is behind the tuned weight by ≥ step. If it lands within a step on any carrier for either method, the control is violated.
EQ2-5 Diagnostic, penalty in other units (SI and MAS, if present in Mammoth at the pinned commit). Same statistic as EQ2-1; expected to miss. Report only; no verdict.
EQ2-6 Ω plateau. On the EWC arm, the best Ω on {0.5, 0.71, 1, 1.41, 2} lies in {0.71, 1, 1.41}. Report the full Ω profile; never write "Ω = 1 exactly", the grid cannot resolve it.
EQ2-7 Published baselines (R7). On the EWC arm, report A-GEM, GradNorm (α = 0, implemented as in its paper with learned weights), and MEGA-I loss-ratio at equal compute. Report only, with one sentence: if any of them is ahead of the rule by ≥ step on every carrier, the rule is dominated by a published method and the row says so.
Any violated control → the mechanism statement in §1 is falsified; the report says so in its first line, whatever EQ2-1 read.
5. Arms, grids, and what to cut if compute is short

Per carrier, in priority order (cut from the bottom, and write the cut into the prereg before hashing):

EWC-online, fixed λ: coarse grid {0.1, 0.3, 1, 3, 10, 30, 100, 300, 1000, 3000, 10000}, five seeds; then a refinement at √2 spacing over [best/√10, best·√10], skipping points already on the coarse grid. The tuned λ is the best of the union. (This two-stage procedure is the pre-registered tuning rule.)
EWC-online, rule at Ω ∈ {0.5, 0.71, 1, 1.41, 2}, five seeds.
EWC-online, reduction arm (fixed w = median derived w), five seeds.
EWC-online sensitivity at Ω = 1: cap ∈ {10, 100, 1e4} × smooth ∈ {0.8, 0.9, 0.98}, five seeds each. A verdict on EQ2-1 that flips in more than one cell is reported "fragile".
ER-sum: fixed w ∈ {0.5, 1, 2, 4}, rule at Ω = 1, reduction arm; five seeds; replay batch = stream batch.
DER++ and LwF: published defaults plus the coarse grid {0.01, 0.03, 0.1, 0.3, 1, 3, 10} on their weight, rule at the five Ω values; five seeds.
A-GEM, GradNorm, MEGA-I on the EWC penalty; five seeds.
SI and MAS: coarse grid plus rule at Ω = 1; five seeds.

Epochs per task, batch size, buffer size (for ER, DER++, A-GEM) and backbone are fixed once for all arms and written into the prereg; equal compute per stream sample is logged (forward/backward counts).

6. Carriers and data hygiene (R11)

Three streams, all absent from [data/SEEN.md](https://github.com/alexsabine/ashes_crr/blob/claude/relaxed-hypatia-m8ikp5/data/SEEN.md) before the hash: Split-CIFAR-100 (10 tasks), Split-TinyImageNet, and a third unseen Mammoth stream chosen by Daniel and named in the prereg. Three are required because EQ2-0 needs a λ spread across carriers and EQ2-1's "every carrier" needs a denominator larger than two. Split-CIFAR-10 and the MNIST family are SEEN and excluded. Record Mammoth's commit hash, dataset versions, download dates and raw-file sha256 in [data/manifests/eq2.sha256](https://github.com/alexsabine/ashes_crr/blob/claude/relaxed-hypatia-m8ikp5/data/manifests/eq2.sha256); append the opened records to [data/SEEN.md](https://github.com/alexsabine/ashes_crr/blob/claude/relaxed-hypatia-m8ikp5/data/SEEN.md) in the same commit as the first download.

7. Order of operations (R2)
uv sync --frozen && uv lock --check
uv run pytest tests
uv run python -m crr.surrogates.gate EQ2 > prereg/eq2/gate_EQ2.txt        # must read OPEN
# write prereg/eq2/PREREG.md from prereg/PREREG_TEMPLATE.md with §2 and §4 verbatim
mkdir -p runs/eq2/frozen && cp <scoring script> src/crr/instrument/core.py theory/CRR.md runs/eq2/frozen/
( cd prereg/eq2 && find . ../../runs/eq2/frozen -type f ! -name '*.pyc' | sort | xargs sha256sum > HASH.txt )
uv run ots stamp prereg/eq2/HASH.txt
git add -A && git commit -m "prereg eq2" && git tag -s prereg-eq2-$(date -I) -m "prereg" && git push --tags
# only now: clone Mammoth at a pinned commit, download the three streams, write the manifest, update SEEN.md
# run once; rerun one unit and cmp; append rows with ledger/append.py; write reports/eq2.md quoting rows only
8. Reporting

[reports/eq2.md](https://github.com/alexsabine/ashes_crr/blob/claude/relaxed-hypatia-m8ikp5/reports/eq2.md) is written after the ledger rows exist and contains no number absent from runs/eq2/. It carries the per-seed table for every arm, the sensitivity table, the exclusion count, and ends with "What a surrogate would have done" pointing at gate_EQ2.txt. Citations (EWC, online EWC, SI, MAS, DER++, LwF, A-GEM, GradNorm, MEGA-I, Mammoth) are fetched on the day with version and date (R10).

9. Outcomes, named now
EQ2-1 PASS, controls 3–4 hold, EQ2-2 "does not reduce": the rule is a tuning-free normalised penalty step for Fisher-type penalties. One narrow, real use. The ledger says exactly that and no more.
EQ2-1 PASS, EQ2-2 "reduces": the rule is a units-fixer that picks λ once per carrier. Also a real use, smaller; say which.
EQ2-1 FAIL: the rule has no use on any existing method. Retire it; the ledger row is the record.
Any control violated: the mechanism is wrong; back to the gate before any further prereg.

**Comment by docxology** [comment]
Phase A executed; **the gate does not open** — recorded in PR #23 with the full evidence.

Summary: three faithful constructions of the S-W positive control (exact Laplace penalty gradient, λ removed, 16× scale mismatch, deterministic seeds) all end GATE CLOSED with S-W failing its MUST_PASS role. The third construction (pure loss-unit mismatch, geometry untouched, present-loss weight swinging 16× with fixed optima) is the decisive one: no collapse, and the rule **reduces to a noisy constant** — the reduction arm ties it exactly (0.0654 = 0.0654) and it loses to the tuned fixed weight by a step (+0.156, 5/5 seeds). Structural reading: for any EWC-type fixed past term a metric-optimal fixed weight exists, and the pull-length ratio carries no cross-task tuning signal beyond transients — unlike the replay family where buffer composition makes the ratio informative (which is why S-V passes gate_EQ).

Per issue §3 and §9 and R12: no prereg written, no hash, no data; the study design rows EQ2-0..7 are untouched. What the run hands back is the Phase-A discovery: **a positive control whose past term's pull-length ratio encodes the tuning demand**. One candidate shape (for Daniel's judgement, not applied): a past term whose curvature is itself state-dependent along the trajectory (so the ratio moves as the optimum is approached) rather than a fixed quadratic — but the gate must be re-run on whatever control is chosen, and the current prereg text stays frozen until then.

R13: the directing prompt and both issue texts are logged verbatim as entry 17 (in PR #23).
